lib/faraday/options.rb

```diff
@@ -48,17 +45,32 @@ def clear
     end
 
     # Public
-    def merge(value)
-      dup.update(value)
+    def merge!(other)
+      other.each do |key, other_value|
+        self_value = self.send(key)
+        sub_options = self.class.options_for(key)
+        if sub_options
+          self_value = sub_options.from(self_value)
```

Comment:
Good point, it should always be a subclass of options at this point.
In fact, we don't even need to `from` the `other_value`, as `each` behaves the same for options and hashes.